jinahub/indexers/query/vector/AnnoyIndexer/__init__.py

```diff
@@ -17,43 +18,66 @@ class AnnoyIndexer(BaseNumpyIndexer):
         :param kwargs:
         """
-        super().__init__(*args, compress_level=0, **kwargs)
+        super().__init__(**kwargs)
+        self.top_k = top_k
         self.metric = metric
-        self.n_trees = n_trees
-        self.search_k = search_k
-
-    def build_advanced_index(self, vecs: 'np.ndarray'):
-        """Build an advanced index structure from a numpy array.
+        self.num_dim = num_dim
+        self.num_trees = num_trees
+        self.indexer = AnnoyIndex(self.num_dim, self.metric)
+        self.traverse_path = traverse_path
+        self.logger = JinaLogger(self.metas.name)
+        dump_path = dump_path or kwargs.get('runtime_args').get('dump_path')
+        if dump_path is not None:
+            self.logger.info('Start building "AnnoyIndexer" from dump data')
+            ids, vecs = import_vectors(dump_path, str(self.metas.pea_id))
+            self._ids = np.array(list(ids))
+            self._vecs = np.array(list(vecs))
+            self._idx_to_ids = {}
+            self._load_index(self._ids, self._vecs)
+        else:
+            self.logger.warning(
+                'No data loaded in "AnnoyIndexer". Use .rolling_update() to re-initialize it...'
+            )
 
-        :param vecs: numpy array containing the vectors to index
-        """
-        from annoy import AnnoyIndex
-        _index = AnnoyIndex(self.num_dim, self.metric)
+    def _load_index(self, ids, vecs):
         for idx, v in enumerate(vecs):
-            _index.add_item(idx, v.astype(np.float32))
-        _index.build(self.n_trees)
-        return _index
-
-    def query(self, keys: 'np.ndarray', top_k: int, *args, **kwargs) -> Tuple['np.ndarray', 'np.ndarray']:
-        """Find the top-k vectors with smallest ``metric`` and return their ids in ascending order.
-        :param keys: numpy array containing vectors to search for
-        :param top_k: upper limit of responses for each search vector
-        """
-        all_idx = []
-        all_dist = []
-        for k in keys:
-            idx, dist = self.query_handler.get_nns_by_vector(k, top_k, self.search_k, include_distances=True)
-            all_idx.append(self._int2ext_id[self.valid_indices][idx])
-            all_dist.append(dist)
-        return np.array(all_idx), np.array(all_dist)
+            self.indexer.add_item(idx, v.astype(np.float32))
+            self._idx_to_ids[idx] = ids[idx]
+        self.indexer.build(self.num_trees)
+
+    @requests(on='/search')
+    def search(self, docs: DocumentArray, **kwargs):
+        for doc in docs.traverse_flat(self.traverse_path):
+            indices, dists = self.indexer.get_nns_by_vector(
+                doc.embedding, self.top_k, include_distances=True
+            )
+            for idx, dist in zip(indices, dists):
+                match = Document(id=self._idx_to_ids[idx], embedding=self._vecs[idx])
+                match.score.value = 1 / (1 + dist)
+                doc.matches.append(match)
+
+    @requests(on='/fill_embedding')
```

Comment:
@JoanFM is this ok?

Reply:
So the idea of this endpoint is that you already have indexed docs, but maybe later you want to fill the same docs with embeddings (probably the embeddings are removed somehow)?

Reply:
the idea of this endpoint is to allow a `recommend` type of interaction (Instead of saying, search me similar items to this UNSEEN documents, find me similar items to this `Id=AAA` document that I already have in my index.

This would involve a 2 way conversation, FILL_EMBEDDING, and Search back with this document